src/group_theory/dihedral.lean

```diff
@@ -0,0 +1,178 @@
+| (sr i) (r j) := sr (i + j)
+| (sr i) (sr j) := r (j - i)
+
+/--
+The identity `1` is the rotation by `0`.
+-/
+def one : dihedral n := r 0
+
+instance : inhabited (dihedral n) := ⟨one⟩
+
+/--
+The inverse of a an element of the dihedral group.
+-/
+def inv : dihedral n → dihedral n
+| (r i) := r (-i)
+| (sr i) := sr i
+
+/--
+The group structure on `dihedral n`.
+-/
+instance : group (dihedral n) :=
+{ mul := mul,
+  mul_assoc :=
+  begin
+    rintros (a | a) (b | b) (c | c);
+    simp only [mul];
+    ring,
+  end,
+  one := one,
+  one_mul :=
+  begin
+    rintros (a | a),
+    exact congr_arg r (zero_add a),
+    exact congr_arg sr (sub_zero a),
+  end,
+  mul_one := begin
+    rintros (a | a),
+    exact congr_arg r (add_zero a),
+    exact congr_arg sr (add_zero a),
+  end,
+  inv := inv,
+  mul_left_inv := begin
+    rintros (a | a),
+    exact congr_arg r (neg_add_self a),
+    exact congr_arg r (sub_self a),
+  end }
+
+lemma r_mul_r (i j : zmod n) : r i * r j = r (i + j) := rfl
+lemma r_mul_sr (i j : zmod n) : r i * sr j = sr (j - i) := rfl
+lemma sr_mul_r (i j : zmod n) : sr i * r j = sr (i + j) := rfl
+lemma sr_mul_sr (i j : zmod n) : sr i * sr j = r (j - i) := rfl
+
+lemma one_def : (1 : dihedral n) = r 0 := rfl
+
+private def fintype_helper : (zmod n ⊕ zmod n) ≃ dihedral n :=
+{ inv_fun := λ i, match i with
+                 | (r j) := sum.inl j
+                 | (sr j) := sum.inr j
+                 end,
+  to_fun := λ i, match i with
+                 | (sum.inl j) := r j
+                 | (sum.inr j) := sr j
+                 end,
+  left_inv := by rintro (x | x); refl,
+  right_inv := by rintro (x | x); refl }
+
+/--
+If `0 < n`, then `dihedral n` is a finite group.
+-/
+instance [fact (0 < n)] : fintype (dihedral n) := fintype.of_equiv _ fintype_helper
+
+instance : nontrivial (dihedral n) := ⟨⟨r 0, sr 0, dec_trivial⟩⟩
+
+/--
+If `0 < n`, then `dihedral n` has `2n` elements.
+-/
+lemma card [fact (0 < n)] : fintype.card (dihedral n) = 2 * n :=
+begin
+  rw ←fintype.card_eq.mpr ⟨fintype_helper⟩,
+  change fintype.card (zmod n ⊕ zmod n) = 2 * n,
+  rw [fintype.card_sum, zmod.card, two_mul]
+end
+
+lemma r_one_pow (k : ℕ) : (r 1 : dihedral n) ^ k = r k :=
+begin
+  induction k with k IH,
+  { refl },
+  { rw [pow_succ, IH, r_mul_r],
+    congr' 1,
+    norm_cast,
+    rw nat.one_add }
+end
+
+lemma r_one_pow_n : (r (1 : zmod n))^n = 1 :=
+begin
+  cases n,
+  { rw pow_zero },
+  { rw [r_one_pow, one_def],
+    congr' 1,
+    exact zmod.cast_self _, }
+end
+
+lemma sr_mul_self (i : zmod n) : sr i * sr i = 1 := by rw [sr_mul_sr, sub_self, one_def]
+
+/--
+If `0 < n`, then `sr i` has order 2.
+-/
+lemma order_of_sr [fact (0 < n)] (i : zmod n) : order_of (sr i) = 2 :=
+begin
+  rw order_of_eq_prime _ _,
+  { exact nat.prime_two },
+  rw [pow_two, sr_mul_self],
+  dec_trivial,
+end
+
+/--
+If `0 < n`, then `r 1` has order `n`.
+-/
+lemma order_of_r_one [hnpos : fact (0 < n)] : order_of (r 1 : dihedral n) = n :=
```

Comment:
```suggestion
@[simp] lemma order_of_r_one [hnpos : fact (0 < n)] : order_of (r 1 : dihedral n) = n :=
```

Reply:
The [linter didn't like this](https://github.com/leanprover-community/mathlib/runs/1555689459) because the left hand side simplifies to `n / n.gcd 1.val`. 

Is there a missing simp lemma for `zmod.val` that would allow us to `simp` `n / n.gcd (1 : zmod n).val` to `n`?

Reply:
For `n ≠ 1`, `(1 : zmod n).val = 1`, so `n / n.gcd (1 : zmod n).val` becomes `n / n.gcd 1 = n / 1 = n`. However, for `n = 1`, `(1 : zmod 1).val = 0, so `n / n.gcd (1 : zmod n).val` becomes `1 / 1.gcd 0 = 1 / 1 = 1`, so I don't see a way for simp to do this without a case split.

Reply:
OK, thanks! This is probably out of scope for this PR anyways.

Reply:
I guess if the lemma below this is un-simped, then this could become a simp-lemma again.